bothan-core/src/registry.rs

```diff
@@ -0,0 +1,121 @@
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use serde::{Deserialize, Serialize};
+
+use crate::registry::signal::Signal;
+use crate::registry::validate::{dfs, ValidationError};
+
+pub mod post_processor;
+pub mod processor;
+pub mod signal;
+pub mod source;
+pub(crate) mod validate;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Invalid;
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Valid;
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Registry<State = Invalid> {
+    #[serde(flatten)]
+    inner: HashMap<String, Signal>,
+    #[serde(skip_serializing, default)]
+    _state: PhantomData<State>,
+}
+
+impl Registry<Invalid> {
+    pub fn validate(self) -> Result<Registry<Valid>, ValidationError> {
+        let mut visited = HashMap::new();
+        for root in self.inner.keys() {
```

Comment:
If you need both keys and values just use `.iter()` (but see another comment below i don't think value is needed)